### PR TITLE
ENH: Add Similarity3DTransform in itkLandmarkBasedTransformInitializer

### PR DESCRIPTION
**Pull request by PranjalSahu**
For adding iso-tropic scaling in the transformation matrix along with Rotation and Translation when using landmarks for transform initialization.
Moved some common code from VersorRigid3DTransform and Similarity3DTransform and created **ComputeCentroid** and **CreateMatrix**.
Introduced a PointType3D to handle the compilation issue when wrapping.

## PR Checklist
- [ ] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [ ] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [X] Added test (or behavior not changed)
- [ ] Updated API documentation (or API not changed)
- [ ] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [ ] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [ ] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKSphinxExamples) for all new major features (if any)

### CODE REVIEW

**Review by thewtex — APPROVED** [review]
:100:

**Review by thewtex — CHANGES_REQUESTED** [review]
2D build error:

  https://open.cdash.org/viewBuildError.php?buildid=8072993

**Comment by PranjalSahu** [comment]
Yes, there is some issue with the Python wrapping caused due to code refactoring.
I am changing the .wrap file to account for it.

**Comment by PranjalSahu** [comment]
Introduced a PointType3D to handle the compilation issue when wrapping.

**Review by dzenanz — APPROVED** [review]
The changes looks fine. Does it compile? 😄

**Comment by PranjalSahu** [comment]
Yes I tested it locally with ctest. Doing one more test with a bigger dataset to check if it gives the correct output. 
With the smaller test sample, it gave the correct scaling output.

**Comment by dzenanz** [comment]
As Matt is on vacation this week, let's not wait for his review. If @blowekamp does not review until 3-4PM, I say we merge.

**Comment by PranjalSahu** [comment]
@dzenanz yes. I discovered one bug in mesh writing in binary mode in rc4.post2 that delayed the testing of this PR.
I will confirm here once that is done.

**Review by blowekamp — APPROVED** [review]
Looks good.

The formatting of the comments could be improved.

File: Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx
Comment (dzenanz) [inline Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx:463]: Reflow this comment paragraph.

File: Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx
Comment (dzenanz) [inline Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx:510]: We don't need this much whitespace in this comment block.

File: Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx
Comment (dzenanz) [inline Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx:538]: I think that ITK style suggest one variable declaration per line.

**Comment by PranjalSahu** [comment]
I have tested it using Python and there it is calling LandmarkBasedTransformInitializer with the VersorRigid3DTransform when using Similarity3DTransform. This is because Similarity3DTransform inherits from VersorRigid3DTransform. I will have to check how to handle this for Python wrappings.

**Comment by PranjalSahu** [comment]
Changing the ordering of dynamic_cast and made it work in Python.

File: Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.h
Comment (jhlegarreta) [inline Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.h:199]: Although there are many other docstrings that do not end with a period (p. 373, 374), at least for the ones that are added/reworked, it'd be nice to add it following the ITK SW Guide.

File: Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx
Comment (jhlegarreta) [inline Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx:365]: For comment blocks, the ITK SW guide recommends using // (p.374), so not sure why this was changed.

File: Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx
Comment (jhlegarreta) [inline Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx:381]: The ITK SW Guide recommends using `this->` to access member methods (p. 333) (applies to any code in the PR; have not checked all calls).

File: Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx
Comment (jhlegarreta) [inline Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx:467]: Not sure this comment is necessary, or whether the else block as a whole is necessary, as it is empty. There is already a mention to the need of having at least 3 landmarks at the beginning of the `if` block.

File: Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx
Comment (jhlegarreta) [inline Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx:511]: Same comment as above.

File: Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx
Comment (jhlegarreta) [inline Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx:556]: Start with capitals (ITK SW Guide).

File: Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx
Comment (jhlegarreta) [inline Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx:605]: Same comment as above about comment block style.

File: Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx
Comment (jhlegarreta) [inline Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx:810]: Start with capitals.

File: Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx
Comment (jhlegarreta) [inline Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx:815]: Same as above.

File: Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx
Comment (jhlegarreta) [inline Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx:420]: Does this "mmm" mean more investigation is required? Can it be tested/checked through a test?

File: Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx
Comment (jhlegarreta) [inline Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx:433]: Stick to the ITK SW Guide comment style guidelines.

**Review by jhlegarreta — COMMENTED** [review]
Some in-line comments, mainly to increase coding style consistency, and one that seems like a suspicious ToDo-style comment.

File: Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx
Comment (PranjalSahu) [reply Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx:420]: I have tested it with a bigger mesh also and it works. I will update the comment.

**Review by thewtex — APPROVED** [review]
:100:

**Pull request merged**
